walkeeper/src/safekeeper.rs

```diff
@@ -408,64 +538,111 @@ where
-    fn handle_vote_request(&mut self, msg: &VoteRequest) -> Result<AcceptorProposerMessage> {
+    fn handle_vote_request(
+        &mut self,
+        msg: &VoteRequest,
+    ) -> Result<Option<AcceptorProposerMessage>> {
         // initialize with refusal
         let mut resp = VoteResponse {
-            term: msg.term,
+            term: self.s.acceptor_state.term,
             vote_given: false as u64,
-            epoch: 0,
             flush_lsn: Lsn(0),
             truncate_lsn: Lsn(0),
+            term_history: TermHistory::empty(),
         };
         if self.s.acceptor_state.term < msg.term {
             self.s.acceptor_state.term = msg.term;
             // persist vote before sending it out
             self.storage.persist(&self.s, true)?;
+            resp.term = self.s.acceptor_state.term;
             resp.vote_given = true as u64;
-            resp.epoch = self.s.acceptor_state.epoch;
             resp.flush_lsn = self.flush_lsn;
             resp.truncate_lsn = self.s.truncate_lsn;
+            resp.term_history = self.get_term_history();
         }
         info!("processed VoteRequest for term {}: {:?}", msg.term, &resp);
-        Ok(AcceptorProposerMessage::VoteResponse(resp))
+        Ok(Some(AcceptorProposerMessage::VoteResponse(resp)))
+    }
+
+    /// Bump our term if received a note from elected proposer with higher one
+    fn bump_if_higher(&mut self, term: Term) -> Result<()> {
+        if self.s.acceptor_state.term < term {
+            self.s.acceptor_state.term = term;
+            self.storage.persist(&self.s, true)?;
+        }
+        Ok(())
+    }
+
+    /// Form AppendResponse from current state.
+    fn append_response(&self) -> AppendResponse {
+        AppendResponse {
+            term: self.s.acceptor_state.term,
+            flush_lsn: self.flush_lsn,
+            commit_lsn: self.s.commit_lsn,
+            disk_consistent_lsn: Lsn(0),
+            // will be filled by the upper code to avoid bothering safekeeper
+            hs_feedback: HotStandbyFeedback::empty(),
+        }
+    }
+
+    fn handle_elected(&mut self, msg: &ProposerElected) -> Result<Option<AcceptorProposerMessage>> {
+        info!("got elected {:?}", msg);
+        self.bump_if_higher(msg.term)?;
+        // If our term is higher, ignore he message (next feedback will inform the compute)
+        if self.s.acceptor_state.term > msg.term {
+            return Ok(None);
+        }
+
+        // TODO: cross check divergence point
+
+        // truncate obsolete part of WAL
+        if self.flush_lsn != Lsn(0) {
+            self.storage
+                .truncate_wal(&self.s.server, msg.start_streaming_at)?;
+        }
+        // update our end of WAL pointer
+        self.flush_lsn = msg.start_streaming_at;
+        // and now adopt term history from proposer
+        self.s.acceptor_state.term_history = msg.term_history.clone();
+        self.storage.persist(&self.s, true)?;
+
+        Ok(None)
```

Comment:
It might be a bit strange, but generally I consider this as a temporary thing as in async proto we'll have different interface/mechanism for sending messages back...